core/domain/stats_services.py

```diff
@@ -22,12 +22,54 @@
 from core.domain import interaction_registry
 from core.domain import stats_domain
 from core.platform import models
+import feconf
 
 (stats_models,) = models.Registry.import_models([models.NAMES.statistics])
 
 # Counts contributions from all versions.
 VERSION_ALL = 'all'
 
+def update_stats(exp_id, exp_version, state_name, event_type, update_params):
+    """Updates ExplorationStatsModel according to the incoming event model.
+
+    Args:
+        exp_id: str. ID of the exploration.
+        exp_version: int. Version of the exploration.
+        state_name: str. Name of the state.
+        event_type: str. Type of the event.
+        update_params: dict. Event specific stats update params.
+    """
+    exploration_stats = get_exploration_stats_by_id(
+        exp_id, exp_version)
+
+    if event_type == feconf.EVENT_TYPE_START_EXPLORATION:
+        exploration_stats.num_starts += 1
+    elif event_type == feconf.EVENT_TYPE_ACTUAL_START_EXPLORATION:
+        exploration_stats.num_actual_starts += 1
+    elif event_type == feconf.EVENT_TYPE_COMPLETE_EXPLORATION:
+        exploration_stats.num_completions += 1
+    elif event_type == feconf.EVENT_TYPE_ANSWER_SUBMITTED:
+        exploration_stats.state_stats_mapping[
+            state_name].total_answers_count += 1
+        if update_params['feedback_is_useful']:
+            exploration_stats.state_stats_mapping[
+                state_name].useful_feedback_count += 1
+    elif event_type == feconf.EVENT_TYPE_STATE_HIT:
+        exploration_stats.state_stats_mapping[
+            state_name].total_hit_count += 1
+        if update_params['is_first_hit']:
+            exploration_stats.state_stats_mapping[
+                state_name].first_hit_count += 1
+    elif event_type == feconf.EVENT_TYPE_STATE_COMPLETED:
+        exploration_stats.state_stats_mapping[
+            state_name].total_completions += 1
```

Comment:
Sorry, just noticed this. But why are some attributes num_X and others total_X or total_X_count? Any reason?

Reply:
Well, everything in ExplorationStats level was num_X whereas most of the StateStats was meant to be total_Y (total_solution_triggered was renamed to num_times_solution_viewed). @seanlip Do they need to be uniform?

Reply:
I think it's weird to have, in particular, total_completions and num_completions, and the fact that you've renamed the "solution" one means you don't really have a naming convention along the lines you describe anymore...

I think semantically "num_X" should be used for transient events, and "total_X_count" should be used for more "thing-like" stuff we're storing somewhere that can be counted -- at least, that seems to be the distinction I'm seeing here. So I would keep total_answers_count, useful_feedback_count, total_hit_count, first_hit_count since answers, feedback, and hits are "things". The only thing I would suggest changing is total_completions to num_completions, since it seems weird to have both num_completions and total_completions.

Reply:
Done.